src/io/flutter/view/FlutterPerfView.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2018 The Chromium Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+package io.flutter.view;
+
+import com.intellij.openapi.Disposable;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.SimpleToolWindowPanel;
+import com.intellij.openapi.util.Disposer;
+import com.intellij.openapi.wm.ToolWindow;
+import com.intellij.openapi.wm.ToolWindowManager;
+import com.intellij.openapi.wm.ex.ToolWindowManagerEx;
+import com.intellij.ui.components.JBLabel;
+import com.intellij.ui.content.Content;
+import com.intellij.ui.content.ContentManager;
+import com.intellij.util.ui.UIUtil;
+import icons.FlutterIcons;
+import io.flutter.run.daemon.FlutterApp;
+import io.flutter.run.daemon.FlutterDevice;
+import io.flutter.utils.VmServiceListenerAdapter;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.*;
+import java.awt.*;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class FlutterPerfView {
+  public static final String TOOL_WINDOW_ID = "Flutter Performance";
+
+  private static final Logger LOG = Logger.getInstance(FlutterPerfView.class);
+
+  @NotNull
+  private final Project myProject;
+
+  private final Map<FlutterApp, PerfViewAppState> perAppViewState = new HashMap<>();
```

Comment:
There does seem to be a lot of duplicate logic managing perAppViewState with FlutterView.
Perhaps add a TabPerAppView<T> to avoid the duplication unless it gets messy?